File: src/model/rice-cooker.ts
import { Elements } from "../constants/elements";
import { Ingredients } from "../constants/ingredients";
import { show } from "../utils/show";

export class RiceCooker {
    public isCooking: boolean = false;
    private temperature: string = "32";
    public hasWater = false;

    private showLoadingAnimation(isReheating: boolean): NodeJS.Timeout {
      const frames = ['(°)', '(¬)', '(°)', '(¬)'];
      let frameIndex = 0;

        const animationInterval = setInterval(() => {
          isReheating ? process.stdout.write(`\r${frames[frameIndex]} En train de réchauffer...`) : process.stdout.write(`\r${frames[frameIndex]} En cours de cuisson...`);
          frameIndex = (frameIndex + 1) % frames.length;
        }, 500);
  
      return animationInterval;
    }

    addWater(): void {
      this.hasWater = true;
    }

    showConfiguration(): void {
      show(`\n La température de globale de cuisson est: ${this.temperature}° \n`)
    } 

    changeConfiguration(newTemp: string): void{
      const convertedTemp = parseInt(newTemp);
      if(isNaN(convertedTemp)){
        show('\n Température invalide \n');
      }else{
        this.temperature = newTemp;
      }
    }

    async startReheating (elementType: string): Promise<void> {
      if((elementType === "ronono") || (elementType === "vary") || (elementType === "laoka")){
        const searchType = Ingredients.find(item => item.name === elementType);
        this.isCooking = true;
        this.hasWater = false
        const loadingAnimation = this.showLoadingAnimation(true);
        const heatingTime = searchType?.time ?? 5

        try {
          await new Promise(resolve => setTimeout(resolve,heatingTime * 1000));
        } catch (error) {
          show('\n Cuisson interrompue \n');
          clearInterval(loadingAnimation);
          this.stopCooking();
        }

        clearInterval(loadingAnimation);

      show('\n Réchauffement terminé \n');

      this.stopCooking();

      }else{
        show("Le rice cooker ne peut pas réchauffer ce type d'élément")
      }
    }
  
    async startCooking (elementType: string, cookingTime: number): Promise<void> {
      if((elementType === "vary gasy") || (elementType === "vary mena") || (elementType === "tsipala")){
        const searchType = Elements.find(item => item.name === elementType);
        this.isCooking = true;
        this.hasWater = false
        const loadingAnimation = this.showLoadingAnimation(false);
        const limite = searchType?.time ?? 5;

        try {
          await new Promise(resolve => setTimeout(resolve, cookingTime * 1000));
        } catch (error) {
          show('\n Cuisson interrompue \n');
          clearInterval(loadingAnimation);
          this.stopCooking();
        }

        clearInterval(loadingAnimation);

      if (limite > cookingTime) {
        show('\n Le riz n\'est pas assez cuit, il manque du temps \n');
      } else if ((cookingTime > limite - 1) && (limite + 2 > cookingTime)) {
        show('\n Le riz est cuit. \n');
      } else {
        show('\n Trop longtemps, le riz est brulé \n');
      }

      this.stopCooking();

      }else{
        show("Le rice cooker ne peut pas cuire ce type d'élément")
      }
    }

    stopCooking(): void {
      if (this.isCooking) {
        this.isCooking = false;
        process.stdin.removeAllListeners('data');
      } else {
        show('\n Le rice cooker n\'est pas en train de cuire. \n');
      }
    }
  }
  

File: src/model/rice_cooker.py
from typing import List
import asyncio

from constants.elements import Elements
from constants.ingredients import Ingredients

class RiceCooker:
    def __init__(self):
        self.isCooking: bool = False
        self.temperature: str = "32"
        self.hasWater: bool = False

    async def showLoadingAnimation(self, isReheating: bool) -> None:
        frames: List[str] = ['(°)', '(¬)', '(°)', '(¬)']
        frameIndex: int = 0

        async def animationCallback() -> None:
            nonlocal frameIndex
            nonlocal isReheating
            while True:
                output = f"\r{frames[frameIndex]} En train de réchauffer..." if isReheating else f"\r{frames[frameIndex]} En cours de cuisson..."
                print(output, end='', flush=True)
                frameIndex = (frameIndex + 1) % len(frames)
                await asyncio.sleep(0.5)

        animation_task = asyncio.create_task(animationCallback())
        return animation_task

    def addWater(self) -> None:
        self.hasWater = True

    def showConfiguration(self) -> None:
        print(f"\n La température de globale de cuisson est: {self.temperature}° \n")

    def changeConfiguration(self, newTemp: str) -> None:
        convertedTemp = int(newTemp) if newTemp.isdigit() else None
        if convertedTemp is None:
            print('\n Température invalide \n')
        else:
            self.temperature = newTemp

    async def startReheating(self, elementType: str) -> None:
        if elementType in {"ronono", "vary", "laoka"}:
            searchType = next((item for item in Ingredients if item.name == elementType), None)
            self.isCooking = True
            self.hasWater = False
            loadingAnimation = await self.showLoadingAnimation(True)
            heatingTime = searchType.time if searchType else 5

            try:
                await asyncio.sleep(heatingTime)
            except Exception as error:
                print('\n Cuisson interrompue \n')
                loadingAnimation.cancel()
                self.stopCooking()

            loadingAnimation.cancel()
            print('\n Réchauffement terminé \n')
            self.stopCooking()

        else:
            print("Le rice cooker ne peut pas réchauffer ce type d'élément")

    async def startCooking(self, elementType: str, cookingTime: int) -> None:
        if elementType in {"vary gasy", "vary mena", "tsipala"}:
            searchType = next((item for item in Elements if item.name == elementType), None)
            self.isCooking = True
            self.hasWater = False
            loadingAnimation = await self.showLoadingAnimation(False)
            limit = searchType.time if searchType else 5

            try:
                await asyncio.sleep(cookingTime)
            except Exception as error:
                print('\n Cuisson interrompue \n')
                loadingAnimation.cancel()
                self.stopCooking()

            loadingAnimation.cancel()

            if limit > cookingTime:
                print('\n Le riz n\'est pas assez cuit, il manque du temps \n')
            elif limit - 1 < cookingTime < limit + 2:
                print('\n Le riz est cuit. \n')
            else:
                print('\n Trop longtemps, le riz est brulé \n')

            self.stopCooking()

        else:
            print("Le rice cooker ne peut pas cuire ce type d'élément")

    def stopCooking(self) -> None:
        if self.isCooking:
            self.isCooking = False
            print('\n Le rice cooker n\'est pas en train de cuire. \n')
        else:
            print('\n Le rice cooker n\'est pas en train de cuire. \n')
